### PR TITLE
feat: Update vcard merge strategies

### PR DESCRIPTION
**Pull request by enguerran**
I added some tests to cover the merge strategies.
I don't know if we may use jest Snapshot or JS expected object.
As I wrote the merge code with TDD, I did not use Jest Snapshot.

### CODE REVIEW

File: src/connections/allContacts.js [hunk 1]
```diff
@@ -104,3 +102,18 @@ export const withContactsMutations = withMutations(client => ({
 }))
 
 export default flow([withContacts, withContactsMutations])
+
+export function mergeContact(contact, attributes) {
+  function concat(arr1 = [], arr2 = []) {
+    return [...arr1, ...arr2]
+  }
+  function customizer(objValue, srcValue, key) {
```
Comment (sebn) [inline src/connections/allContacts.js]: What about making it top-level, so we could unit-test attribute merging logic on its own without having to deal with a full contact object in every test?

Although some people consider that *private* functions should not be exposed for testing purpose, maybe it's worth the trouble.

Array attributes would probably still need to be tested as a whole in order to be able to either merge two corresponding items or keep them separate depending on their content.

File: src/connections/allContacts.js [hunk 2]
```diff
@@ -104,3 +102,18 @@ export const withContactsMutations = withMutations(client => ({
...
+    if (Array.isArray(objValue) || Array.isArray(srcValue)) {
+      return concat(objValue, srcValue)
```
Comment (sebn) [inline src/connections/allContacts.js]: We'll probably want to prevent duplicates at some point, but I assume this will happen in another PR :slightly_smiling_face:

File: src/connections/allContacts.js [hunk 3]
```diff
@@ -104,3 +102,18 @@ export const withContactsMutations = withMutations(client => ({
...
+    }
+    if (key === 'fullname') {
+      return objValue.length > srcValue.length ? objValue : srcValue
```
Comment (sebn) [inline src/connections/allContacts.js]: Wouldn't this make data loss possible?
E.g. `{fullname: "John J. Doe", ...}` and `{fullname: "John Doe Jr", ...}`? (sorry, I can't think of a better example)

What about checking whether some of the strings include the other one, defaulting to both slash-concatenated? It should work assuming the attributes containers (contacts here) are considered identicals (i.e. with the same identity, not necessarily the same value).

File: src/connections/allContacts.js [hunk 4]
```diff
@@ -104,3 +102,18 @@ export const withContactsMutations = withMutations(client => ({
...
+    }
+  }
+  return mergeWith({ ...contact }, attributes, customizer)
```
Comment (sebn) [inline src/connections/allContacts.js]: :+1: `mergeWith` looks like a good solution!

File: src/connections/allContacts.js [hunk 5]
```diff
@@ -104,3 +102,18 @@ export const withContactsMutations = withMutations(client => ({
 }))
 
 export default flow([withContacts, withContactsMutations])
+
+export function mergeContact(contact, attributes) {
+  function concat(arr1 = [], arr2 = []) {
+    return [...arr1, ...arr2]
```
Comment (sebn) [inline src/connections/allContacts.js]: :ok_hand: It's weird that neither `Array#concat()`, `Array.from()`, `_.concat()` nor the spread operator handle a `null` value as an empty sequence :confused:

File: src/connections/allContacts.js [hunk 6]
```diff
@@ -104,3 +102,18 @@ export const withContactsMutations = withMutations(client => ({
...
+  }
+  function customizer(objValue, srcValue, key) {
+    if (Array.isArray(objValue) || Array.isArray(srcValue)) {
+      return concat(objValue, srcValue)
+    }
+    if (key === 'fullname') {
```
Comment (sebn) [inline src/connections/allContacts.js]: Couldn't this be generalized to other string attributes? (I'm not sure, just asking)

Also what happens to them with the current implementation?

**Comment by sebn** [comment]
> I don't know if we may use jest Snapshot or JS expected object.
> As I wrote the merge code with TDD, I did not use Jest Snapshot.

In my initial implementation, I used both:
- Snapshots in integration tests to have full output examples and prevent regressions (but I finally excluded them from the PR according to feedback)
- Unit tests for (almost) every function, testing edge cases at the lowest possible level

Should I have to do it again, I believe I'd go the same way.

File: src/connections/allContacts.js (hunk 3 above)
Comment (sebn) [reply src/connections/allContacts.js]: (this one sound like a blocker)

File: src/connections/allContacts.js (hunk 6 above)
Comment (sebn) [reply src/connections/allContacts.js]: Actually I don't remember of other top-level string attributes... Are there any?

**Comment by sebn** [comment]
Also maybe we could introduce some merge idempotence testing (i.e. making sure that merging the same contacts twice ends up with the same result)?

File: src/connections/allContacts.js (hunk 2 above)
Comment (enguerran) [reply src/connections/allContacts.js]: Ohhh, you are so right.
No, I want to add that to this PR.

File: src/connections/allContacts.js (hunk 5 above)
Comment (enguerran) [reply src/connections/allContacts.js]: Ahhh EcmaScript :disappointed:
:rofl:

File: src/connections/allContacts.js (hunk 1 above)
Comment (enguerran) [reply src/connections/allContacts.js]: I understand your proposal, however:
The added value of the functionality is the merge of contacts, a unit test on an implementation detail does not seem easy to maintain.

Comment (enguerran) [reply src/connections/allContacts.js]: @sebn I refactored the way I manage merge strategies. Can you take a look at it?

File: src/connections/allContacts.js (hunk 3 above)
Comment (enguerran) [reply src/connections/allContacts.js]: In https://github.com/cozy/cozy-doctypes/pull/57 I suggest to remove this field.

And I changed the target of `takeLonger` strategy to become `givenName` and `familyName`.

**Comment by enguerran** [comment]
@sebn I refactored the way I manage merge strategies. Can you take a look at it?

File: src/connections/allContacts.js [hunk 7]
```diff
@@ -104,3 +105,45 @@ export const withContactsMutations = withMutations(client => ({
...
+export function mergeContact(contact, attributes) {
+  // TODO: refactor to remove customizer from the mergeContact
+  const propertyName = {
+    email: 'address',
+    phone: 'number'
+  }
+
+  function customizer(objValue, srcValue, key) {
+    if (Array.isArray(objValue) || Array.isArray(srcValue)) {
+      return mergeArrayOnKey(objValue, srcValue, propertyName[key])
```
Comment (y-lohse) [inline src/connections/allContacts.js]: What happens if `propertyName[key]` is `undefined`? Can that happen?

Comment (enguerran) [reply src/connections/allContacts.js]: I changed the implementation, can you check if it is all good.
If there is no strategie, the classical lodash's merge is used.

**Comment by enguerran** [comment]
@y-lohse @sebn can you check if the new implementation looks good to you. Regarding the tests, they cover the cases in mind when implementing the features. If we want some additional check, we juste have to add a test and implement in TDD mode.

File: src/helpers/mergeContact.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  email: mergeArrayOfObject('address'),
+  phone: mergeArrayOfObject('number'),
+  givenName: keepLonger,
+  familyName: keepLonger
+}
+
+const fromStrategies = strategies => (objValue, srcValue, key) =>
+  strategies[key] ? strategies[key](objValue, srcValue, key) : undefined
+
+const customizer = fromStrategies(strategies)
```
[the remark this answers is not in the document]
Comment (enguerran) [reply src/helpers/mergeContact.js]: It is written for readability and understandability. The named function that takes an object `strategies` and returns a function with 3 parameters `objValue, srcValue, key` and returns `undefined` or a property of the object `strategies` express its intention with this naming.

File: src/helpers/mergeContact.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+function mergeArrayOfObject(key) {
+  return (objValue, srcValue) => {
+    return mergeArrayOnKey(objValue, srcValue, key)
+  }
+}
+
+function keepLonger(objValue = '', srcValue = '') {
+  return objValue.length > srcValue.length ? objValue : srcValue
```
[the remark this answers is not in the document]
Comment (enguerran) [reply src/helpers/mergeContact.js]: We use the length of the string as a basis for the strategy that wants to keep the longest string.

The Product Owner wrote a story that says the user wants to keep the longest chain because it is a good heuristic to keep the richest information.
Unless unrealistic examples such as `"John''` vs `"AAAAAAAAAAAAA''`, we can consider that `"John''` contains less information than `"Lord Commander John''` or `"Jonathan''`.

File: src/helpers/mergeContact.spec.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    const fakeContact = {
+      email: [{ address: 'starck@localhost' }],
+      name: { familyName: 'Starck' },
+      phone: [{ number: '0987654321', primary: true, type: 'Work' }],
+      address: [{ formattedAddress: 'Westeros', primary: true, type: 'Work' }]
+    }
+    const importedContact = {
+      name: { familyName: 'Starck', givenName: 'Ned' },
+      phone: [{ number: '0123456789' }],
+      email: [{ address: 'ned@localhost' }]
```
[the remark this answers is not in the document]
Comment (enguerran) [reply src/helpers/mergeContact.spec.js]: In this test, we are given two contacts and check the merge with the expected merged contact. We don't decide if the contact should be merged or not. But you are right, we could add a property common to both contacts, but it will not change the test expectation.

File: src/connections/allContacts.js [hunk 11]
```diff
@@ -62,11 +16,7 @@ export const withContactsMutations = withMutations(client => ({
...
     if (contacts.length === 1) {
-      // TODO: create a more complex updated contact if we need to merge deep properties together.
-      const updatedContact = {
-        ...contacts[0],
-        ...attributes
-      }
+      const updatedContact = mergeContact(contacts[0], attributes)
       return client.save(updatedContact)
     } else if (contacts.length > 1) {
       throw new Error(
```
[the remark this answers is not in the document]
Comment (enguerran) [reply src/connections/allContacts.js]: I guess someday the application will have a merging tool.
